### DISCUSSION TITLE
Can HLS deal, in a given IDE session, with files belonging to different projects?

### DISCUSSION THREAD

**Discussion by Aster89** [haskell/haskell-language-server#4048] [Q&A]
I manage my Haskell development via GHCup, so I use `ghcup tui` to install and update `cabal`/GHC/HLS (no Stack installed), and `cabal` to install libraries if I need them (e.g. `cabal install --lib flow`, if I needed to `import Flow`).

Assume I have two sibling directories as these,
```none
.
├── dir1
│   ├── bar.cabal
│   ├── bar.hs
│   └── cabal.project
└── dir2
    ├── cabal.project
    ├── foo.cabal
    └── foo.hs
```
where `foo.hs` contains
```haskell
import Data.Text
main :: IO ()
main = putStrLn "hello"
```
`bar.hs` contains
```haskell
main :: IO ()
main = putStrLn "hello"
```
`foo.cabal` contains
```cabal
cabal-version:       3.0
name:                foo
version:             0.1.0.0
executable foo
  main-is:             foo.hs
  build-depends:       base >= 4.13
                     , text
```
`bar.cabal`
```cabal
cabal-version:       3.0
name:                bar
version:             0.1.0.0
executable bar
  main-is:             bar.hs
  build-depends:       base >= 4.13
```
and both `cabal.project` files contain
```cabal
packages: .
with-compiler: ghc-9.4.8
package foo
  flags: +all_extensions
```

In Vim (I use YouCompleteMe plugin) I can verify that HLS is running in the project directory of the first of those two files to be open; indeed, if I opened Vim via `vim -O dir1/bar.hs dir2/foo.hs` (which opens `bar.hs` on the left and focuses it first, and `foo.hs` on the right) all is fine with `bar.hs`, but as soon as I moving to the other file, an error like the following is shown
```
Loading the module '/home/enrico/deleteme/deleteme/dir2/foo.hs' failed.

It may not be listed in your .cabal file!
Perhaps you need to add `foo` to other-modules or exposed-modules.

For more information, visit: https://cabal.readthedocs.io/en/3.4/developing-packages.html#modules-included-in-the-package

"././bar.hs" - Cabal {component = Just "bar:exe:bar"}
```
which seems to support my understanding that HLS has detected one project and got "stuck" on it.

If I delete the two `cabal.project` files, so the structure is
```
.
├── dir1
│   ├── bar.cabal
│   └── bar.hs
└── dir2
    ├── foo.cabal
    └── foo.hs
```
the issue does not occur, so I think the problem occurs when jumping across two projects.

Is this the case? Or, if the IDE that uses HLS is supposed to handle the scenario above, what information can I provide for you to help me understand what I'm doing wrong?

I've already spoken to YouCompleteMe devs, and they say that Vim+YCM seems to be sending the correct info to HLS, so I'm coming here to seek some help understanding what I'm doing wrong.

Here is a screencast: [untitled](https://asciinema.org/a/O65ZV3uGibcNQjWvdBjbQhOXr) - asciinema.org

**Comment by michaelpj** [comment]
HLS can cope with multiple or components within one cabal project, but does not work across multiple projects. You need to configure your editor to consider both of those to be "workspace folders", which will probably result in an independent HLS process for each.
